working/0723-static-extensions/feature-specification-variant2.md

```diff
@@ -0,0 +1,347 @@
+
+A static member invocation on a class `C`, of the form `C.m()` (or any
+other member access), is resolved by looking up static members in `C` named
+`m` and looking up static members of every accessible extension with
+on-class `C` and a static member named `m`.
+
+If `C` contains such a declaration then the expression is an invocation of
+that static member of `C`, with the same static analysis and dynamic
+behavior as before the introduction of this feature.
+
+Otherwise, an error occurs if fewer than one or more than one declaration
+named `m` was found. *They would necessarily be declared in extensions.*
+
+Otherwise, the invocation is resolved to the given static member
+declaration in an extension named `Ej`, and the invocation is treated
+as `Ej.m()` *(this is an explicitly resolved invocation, which is specified
+above)*.
+
+#### The instantiated constructor return type of an extension
+
+We associate an extension declaration _D_ named `E` with formal type
+parameters `X1 extends B1 .. Xs extends Bs` and an actual type argument
+list `T1 .. Ts` with a type known as the _instantiated constructor return
+type of_ _D_ _with type arguments_ `T1 .. Ts`.
+
+When an extension declaration _D_ named `E` has an on-clause which is
+a non-generic class `C`, the instantiated constructor return type is `C`,
+for any list of actual type arguments.
+
+*It is not very useful to declare a type parameter of an extension
+which isn't used in the constructor return type, because it can only be
+passed in an explicitly resolved constructor invocation, e.g.,
+`E<int>.C(42)`. In all other invocations, the value of such type variables
+is determined by instantiation to bound. In any case, the type parameters
+are always ignored by static member declarations, they are only relevant to
+constructors.*
+
+When an extension declaration _D_ has no formal type parameters, and
+it has an on-type `C<S1 .. Sk>`, the instantiated constructor return type
+of _D_ is `C<S1 .. Sk>`. *In this case the on-type is a fixed type (also
+known as a ground type), e.g., `List<int>`. This implies that the
+constructor return type of D is the same for every call site.*
+
+Finally we have the general case: Consider an extension declaration
+_D_ named `E` with formal type parameters `X1 extends B1 .. Xs extends Bs`
+and a constructor return type `C<S1 .. Sk>`. With actual type arguments
+`T1 .. Ts`, the instantiated constructor return type of _D_ with type
+arguments `T1 .. Ts` is `[T1/X1 .. Ts/Xs]C<S1 .. Sk>`.
+
+#### Explicit invocation of a constructor in an extension
+
+Explicit constructor invocations are similar to static member invocations,
+but they need more detailed rules because they can use the formal type
+parameters declared by an extension.
+
+An _explicitly resolved invocation_ of a constructor named `C.name` in a
+extension declaration _D_ named `E` with `s` type parameters and on-class
+`C` can be expressed as `E<S1 .. Ss>.C.name(args)`, 
```

Comment:
It is `E<S1 .. Ss>.C.name(args)` because we need to be able to check all parts against each other (arriving at `E<S1 .. Ss>.C<U1 .. Uk>.name(args)` somehow). This is what it takes to specify a fully explicit static-extension constructor invocation. It is redundant but easy to check, all other forms will be transformed into this fully explicit form in order to define the meaning of the less informative versions.

Reply:
It might not be new syntax since `<S1...,Sn>` was made a selector, but I think it does need its own "Grammar-interpretation" section to describe how it's interpreted. It's a new syntax combination, previously a constructor invocation could not have the form `id<Types>.id.id(args)`. 
Something else could, it would not be a single constructor invocation.

Reply:
> This is what it takes to specify a fully explicit static-extension constructor invocation.

Only if we *want* that syntax. (Or we *need* it, which may be the case.)

Otherwise I'd have gone with `extension ... on C<...> { E.name(params) ... }`, where `E.name` is a constructor which can also be accessed as `C<T1,..,Tn>.name`, with inference going from `T1`...`Tn` to the type parameters of `N` through the `on` type.

I think that can work for invocation, but it breaks down for tear-offs. Maybe it breaks down if you abstract over the types used in the invocation too.

I can see why `C.name` would make sense, making it a constructor that has the same type variables of `C`, but then I'd want to write those type parameters on the constructor myself, not inherit them from somewhere else.
That is:
```dart
extension E<K, V> on Map<K, V> {
  factory Map<K extends Comparable<K>, V>.ordered() => SplayTreeMap<K, V>(compareComparable<K>);
}
```
Which is new syntax.
Without new syntax, it can at most be:
```dart
extension E<K2, V2> on Map<K2, V2> {
  factory Map.ordered() => SplayTreeMap<K, V>(compareComparable<K>);
}
```
where it's very unclear where the type variables available to `Map.ordered` comes from. If anywhere.